/*
NaiveSystems Analyze - A tool for static code analysis
Copyright (C) 2023  Naive Systems Ltd.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

syntax = "proto3";
package analyzer.proto;

option go_package = "analyzer/proto";

// An error location as part of a multi-location analysis result.
message ErrorLocation {
  // Path of the analyzed file relative to the project root directory.
  string path = 1;
  // Line number in the analyzed file where the result is attached.
  int32 line_number = 2;
}

// An analysis result.
message Result {
  // Absolute Path of the analyzed file.
  string path = 1;
  // Line number in the analyzed file where the result is attached.
  int32 line_number = 2;
  // A list of locations related to the same result.
  // For example, the analyzer may output multiple locations for MISRA
  // system checks.
  repeated ErrorLocation locations = 4;
  // A human-readable representation of the error message.
  string error_message = 3;
  // A flag representing false positive given by specific checkers.
  bool false_positive = 5;
  // For structured error messages.
  enum ErrorKind {
    UNKNOWN = 0;
    // prefix 1 for misra-c-2012 rule
    // postfix 00 for cppcheck default situation
    MISRA_C_2012_RULE_1_1_STRUCT_MEMBER = 1010101;
    MISRA_C_2012_RULE_1_1_FUNCTION_PARM = 1010102;
    MISRA_C_2012_RULE_1_1_FUNCTION_ARG = 1010103;
    MISRA_C_2012_RULE_1_1_NESTED_RECORD = 1010104;
    MISRA_C_2012_RULE_1_1_NESTED_EXPR = 1010105;
    MISRA_C_2012_RULE_1_1_SWITCH_CASE = 1010106;
    MISRA_C_2012_RULE_1_1_ENUM_CONSTANT = 1010107;
    MISRA_C_2012_RULE_1_1_STRING_CHAR = 1010108;
    MISRA_C_2012_RULE_1_1_EXTERN_ID = 1010109;
    MISRA_C_2012_RULE_1_1_EXTERN_ID_CHAR = 1010110;
    MISRA_C_2012_RULE_1_1_MACRO_ID = 1010111;
    MISRA_C_2012_RULE_1_1_MACRO_PARM = 1010112;
    MISRA_C_2012_RULE_1_1_MACRO_ARG = 1010113;
    MISRA_C_2012_RULE_1_1_NESTED_BLOCK = 1010114;
    MISRA_C_2012_RULE_1_1_NESTED_INCLUDE = 1010115;
    MISRA_C_2012_RULE_1_1_IOM_ID_CHAR = 1010116;
    MISRA_C_2012_RULE_1_1_NESTED_COND_INCLU = 1010117;
    MISRA_C_2012_RULE_1_1_BLOCK_ID = 1010118;
    MISRA_C_2012_RULE_1_1_NESTED_DECL = 1010119;
    MISRA_C_2012_RULE_1_1_MODIFY_DECL = 1010120;
    MISRA_C_2012_RULE_1_3 = 1010301;
    MISRA_C_2012_RULE_1_4 = 1010400;
    MISRA_C_2012_RULE_2_1 = 1020101;
    MISRA_C_2012_RULE_2_2 = 1020201;
    MISRA_C_2012_RULE_2_2_EXTERNAL = 1020202;
    MISRA_C_2012_RULE_2_3 = 1020301;
    MISRA_C_2012_RULE_2_4 = 1020400;
    MISRA_C_2012_RULE_2_5 = 1020500;
    MISRA_C_2012_RULE_2_6 = 1020600;
    MISRA_C_2012_RULE_2_7 = 1020700;
    MISRA_C_2012_RULE_3_1 = 1030100;
    MISRA_C_2012_RULE_3_2 = 1030200;
    MISRA_C_2012_RULE_4_1 = 1040100;
    MISRA_C_2012_RULE_4_2 = 1040200;
    MISRA_C_2012_RULE_5_1_NON_ASCII_ERROR = 1050101;
    MISRA_C_2012_RULE_5_1_DISTINCT_ERROR = 1050102;
    MISRA_C_2012_RULE_5_2 = 1050200;
    MISRA_C_2012_RULE_5_3 = 1050300;
    MISRA_C_2012_RULE_5_4 = 1050400;
    MISRA_C_2012_RULE_5_5 = 1050500;
    MISRA_C_2012_RULE_5_6 = 1050601;
    MISRA_C_2012_RULE_5_7_DUPLICATE = 1050701;
    MISRA_C_2012_RULE_5_7_INVALID = 1050702;
    MISRA_C_2012_RULE_5_8 = 1050801;
    MISRA_C_2012_RULE_5_9 = 1050901;
    MISRA_C_2012_RULE_6_1 = 1060100;
    MISRA_C_2012_RULE_6_2 = 1060200;
    MISRA_C_2012_RULE_7_1 = 1070100;
    MISRA_C_2012_RULE_7_2 = 1070200;
    MISRA_C_2012_RULE_7_3 = 1070300;
    MISRA_C_2012_RULE_7_4_CPPCHECK = 1070400;
    MISRA_C_2012_RULE_7_4 = 1070401;
    MISRA_C_2012_RULE_7_4_EXTERNAL = 1070402;
    MISRA_C_2012_RULE_8_1 = 1080100;
    MISRA_C_2012_RULE_8_2_FUNC_DECL_PARM_NOT_NAMED_ERROR = 1080201;
    MISRA_C_2012_RULE_8_2_FUNC_POINTER_PARM_NOT_NAMED_ERROR = 1080202;
    MISRA_C_2012_RULE_8_2_FUNC_DECL_KR_STYLE_ERROR = 1080203;
    MISRA_C_2012_RULE_8_2_FUNC_DECL_VOID_ERROR = 1080204;
    MISRA_C_2012_RULE_8_3_INVALID_DECL_ERROR = 1080301;
    MISRA_C_2012_RULE_8_3_ERROR = 1080302;
    MISRA_C_2012_RULE_8_4_NO_FUNC_DECL_ERROR = 1080401;
    MISRA_C_2012_RULE_8_4_NO_VAR_DECL_ERROR = 1080402;
    MISRA_C_2012_RULE_8_4_FUNC_PARM_NOT_MATCH_ERROR = 1080403;
    MISRA_C_2012_RULE_8_5_EXT_VD_IN_C_ERROR = 1080501;
    MISRA_C_2012_RULE_8_5_EXT_FD_IN_C_ERROR = 1080502;
    MISRA_C_2012_RULE_8_5_EXT_VD_DUP_ERROR = 1080503;
    MISRA_C_2012_RULE_8_5_EXT_FD_DUP_ERROR = 1080504;
    MISRA_C_2012_RULE_8_6 = 1080601;
    MISRA_C_2012_RULE_8_7 = 1080701;
    MISRA_C_2012_RULE_8_8 = 1080800;
    MISRA_C_2012_RULE_8_9 = 1080900;
    MISRA_C_2012_RULE_8_10 = 1081000;
    MISRA_C_2012_RULE_8_11 = 1081100;
    MISRA_C_2012_RULE_8_12 = 1081200;
    MISRA_C_2012_RULE_8_14 = 1081400;
    MISRA_C_2012_RULE_9_1 = 1090101;
    MISRA_C_2012_RULE_9_2 = 1090200;
    MISRA_C_2012_RULE_9_3 = 1090300;
    MISRA_C_2012_RULE_9_4 = 1090400;
    MISRA_C_2012_RULE_9_5 = 1090500;
    MISRA_C_2012_RULE_10_1 = 1100100;
    MISRA_C_2012_RULE_10_2 = 1100200;
    MISRA_C_2012_RULE_10_3 = 1100300;
    MISRA_C_2012_RULE_10_4 = 1100400;
    MISRA_C_2012_RULE_10_5 = 1100500;
    MISRA_C_2012_RULE_10_6 = 1100600;
    MISRA_C_2012_RULE_10_7 = 1100700;
    MISRA_C_2012_RULE_10_8 = 1100800;
    MISRA_C_2012_RULE_11_1 = 1110101;
    MISRA_C_2012_RULE_11_1_EXTERNAL = 1110102;
    MISRA_C_2012_RULE_11_2 = 1110201;
    MISRA_C_2012_RULE_11_3 = 1110301;
    MISRA_C_2012_RULE_11_4_CPPCHECK = 1110400;
    MISRA_C_2012_RULE_11_4_INT_TO_POINTER = 1110401;
    MISRA_C_2012_RULE_11_4 = 1110402;
    MISRA_C_2012_RULE_11_5_CPPCHECK = 1110500;
    MISRA_C_2012_RULE_11_5 = 1110501;
    MISRA_C_2012_RULE_11_6 = 1110601;
    MISRA_C_2012_RULE_11_6_EXTERNAL = 1110602;
    MISRA_C_2012_RULE_11_7 = 1110701;
    MISRA_C_2012_RULE_11_7_EXTERNAL = 1110702;
    MISRA_C_2012_RULE_11_8 = 1110801;
    MISRA_C_2012_RULE_11_9 = 1110900;
    MISRA_C_2012_RULE_12_1 = 1120100;
    MISRA_C_2012_RULE_12_2 = 1120201;
    MISRA_C_2012_RULE_12_3 = 1120301;
    MISRA_C_2012_RULE_12_4 = 1120400;
    MISRA_C_2012_RULE_12_5 = 1120500;
    MISRA_C_2012_RULE_13_1 = 1130101;
    MISRA_C_2012_RULE_13_2 = 1130201;
    MISRA_C_2012_RULE_13_2_EXTERNAL = 1130202;
    MISRA_C_2012_RULE_13_3 = 1130301;
    MISRA_C_2012_RULE_13_4 = 1130401;
    MISRA_C_2012_RULE_13_4_EXTERNAL = 1130402;
    MISRA_C_2012_RULE_13_5 = 1130501;
    MISRA_C_2012_RULE_13_5_IMPURE_FUNCTION = 1130502;
    MISRA_C_2012_RULE_13_6 = 1130600;
    MISRA_C_2012_RULE_14_1 = 1140101;
    MISRA_C_2012_RULE_14_1_CSA = 1140102;
    MISRA_C_2012_RULE_14_2_MAY_HAVE_PSE = 1140201;
    MISRA_C_2012_RULE_14_2_COUNTER_NOT_USED = 1140202;
    MISRA_C_2012_RULE_14_2_COUNTER_MODIFIED_IN_BODY = 1140203;
    MISRA_C_2012_RULE_14_2_REF_MODIFIED_IN_BODY = 1140204;
    MISRA_C_2012_RULE_14_2_CLAUSE_SET_OTHER_VALUE = 1140205;
    MISRA_C_2012_RULE_14_2_2ND_CLAUSE_USE_NON_BOOL = 1140206;
    MISRA_C_2012_RULE_14_3 = 1140301;
    MISRA_C_2012_RULE_14_3_EXTERNAL = 1140302;
    MISRA_C_2012_RULE_14_4 = 1140400;
    MISRA_C_2012_RULE_15_1 = 1150100;
    MISRA_C_2012_RULE_15_2 = 1150200;
    MISRA_C_2012_RULE_15_3 = 1150300;
    MISRA_C_2012_RULE_15_4 = 1150400;
    MISRA_C_2012_RULE_15_5 = 1150500;
    MISRA_C_2012_RULE_15_6 = 1150600;
    MISRA_C_2012_RULE_15_7 = 1150700;
    MISRA_C_2012_RULE_16_1 = 1160100;
    MISRA_C_2012_RULE_16_2 = 1160200;
    MISRA_C_2012_RULE_16_3 = 1160300;
    MISRA_C_2012_RULE_16_4 = 1160400;
    MISRA_C_2012_RULE_16_5 = 1160500;
    MISRA_C_2012_RULE_16_6 = 1160600;
    MISRA_C_2012_RULE_16_7 = 1160700;
    MISRA_C_2012_RULE_17_1 = 1170100;
    MISRA_C_2012_RULE_17_2 = 1170201;
    MISRA_C_2012_RULE_17_2_EXTERNAL = 1170202;
    MISRA_C_2012_RULE_17_3 = 1170301;
    MISRA_C_2012_RULE_17_4_CPPCHECK = 1170400;
    MISRA_C_2012_RULE_17_4 = 1170401;
    MISRA_C_2012_RULE_17_5_NULL_POINTER_ERROR = 1170501;
    MISRA_C_2012_RULE_17_5_ARRAY_SIZE_ERROR = 1170502;
    MISRA_C_2012_RULE_17_5_EXTERNAL = 1170503;
    MISRA_C_2012_RULE_17_6 = 1170600;
    MISRA_C_2012_RULE_17_7 = 1170700;
    MISRA_C_2012_RULE_17_8_CPPCHECK = 1170800;
    MISRA_C_2012_RULE_17_8 = 1170801;
    MISRA_C_2012_RULE_18_1 = 1180101;
    MISRA_C_2012_RULE_18_2 = 1180201;
    MISRA_C_2012_RULE_18_3 = 1180301;
    MISRA_C_2012_RULE_18_4 = 1180400;
    MISRA_C_2012_RULE_18_5 = 1180500;
    MISRA_C_2012_RULE_18_6 = 1180601;
    MISRA_C_2012_RULE_18_7 = 1180700;
    MISRA_C_2012_RULE_18_8 = 1180800;
    MISRA_C_2012_RULE_19_1 = 1190101;
    MISRA_C_2012_RULE_19_1_EXTERNAL = 1190102;
    MISRA_C_2012_RULE_19_2 = 1190200;
    MISRA_C_2012_RULE_20_1 = 1200100;
    MISRA_C_2012_RULE_20_2 = 1200200;
    MISRA_C_2012_RULE_20_3 = 1200300;
    MISRA_C_2012_RULE_20_4 = 1200400;
    MISRA_C_2012_RULE_20_5 = 1200500;
    MISRA_C_2012_RULE_20_6 = 1200600;
    MISRA_C_2012_RULE_20_7 = 1200700;
    MISRA_C_2012_RULE_20_8 = 1200800;
    MISRA_C_2012_RULE_20_9 = 1200900;
    MISRA_C_2012_RULE_20_10 = 1201000;
    MISRA_C_2012_RULE_20_11 = 1201100;
    MISRA_C_2012_RULE_20_12 = 1201200;
    MISRA_C_2012_RULE_20_13 = 1201300;
    MISRA_C_2012_RULE_20_14 = 1201400;
    MISRA_C_2012_RULE_21_1 = 1210100;
    MISRA_C_2012_RULE_21_2 = 1210200;
    MISRA_C_2012_RULE_21_3 = 1210300;
    MISRA_C_2012_RULE_21_4 = 1210400;
    MISRA_C_2012_RULE_21_5 = 1210500;
    MISRA_C_2012_RULE_21_6 = 1210600;
    MISRA_C_2012_RULE_21_7 = 1210700;
    MISRA_C_2012_RULE_21_8 = 1210800;
    MISRA_C_2012_RULE_21_9 = 1210900;
    MISRA_C_2012_RULE_21_10 = 1211000;
    MISRA_C_2012_RULE_21_11 = 1211100;
    MISRA_C_2012_RULE_21_12 = 1211200;
    MISRA_C_2012_RULE_21_13 = 1211301;
    MISRA_C_2012_RULE_21_14 = 1211401;
    MISRA_C_2012_RULE_21_15 = 1211500;
    MISRA_C_2012_RULE_21_16 = 1211600;
    MISRA_C_2012_RULE_21_17 = 1211701;
    MISRA_C_2012_RULE_21_18 = 1211801;
    MISRA_C_2012_RULE_21_18_EXTERNAL = 1211802;
    MISRA_C_2012_RULE_21_19 = 1211901;
    MISRA_C_2012_RULE_21_19_EXTERNAL = 1211902;
    MISRA_C_2012_RULE_21_20 = 1212001;
    MISRA_C_2012_RULE_21_21 = 1212100;
    MISRA_C_2012_RULE_22_1 = 1220101;
    MISRA_C_2012_RULE_22_2 = 1220201;
    MISRA_C_2012_RULE_22_3 = 1220301;
    MISRA_C_2012_RULE_22_4 = 1220401;
    MISRA_C_2012_RULE_22_5 = 1220501;
    MISRA_C_2012_RULE_22_6 = 1220601;
    MISRA_C_2012_RULE_22_7 = 1220701;
    MISRA_C_2012_RULE_22_8 = 1220801;
    MISRA_C_2012_RULE_22_9 = 1220901;
    MISRA_C_2012_RULE_22_10 = 1221001;
    // prefix 2 for misra-c-2012 dir
    MISRA_C_2012_DIR_4_3_ASM_SHOULD_BE_ISOLATED = 2040301;
    MISRA_C_2012_DIR_4_3_ASM_SHOULD_BE_ENCAPSULATED = 2040302;
    MISRA_C_2012_DIR_4_7_ERROR_SHOULD_BE_TESTED = 2040701;
    MISRA_C_2012_DIR_4_10_HAS_NO_PRECAUTION = 2041001;
    MISRA_C_2012_DIR_4_10_HAS_SAME_FILE_ID = 2041002;
    MISRA_C_2012_DIR_4_11 = 2041101;
    MISRA_C_2012_DIR_4_12 = 2041201;
    MISRA_C_2012_DIR_4_14 = 2041401;
    // prefix 3 for self-defined rules
    NAIVESYSTEMS_C7966 = 3796601;
    // prefix 4 for misra-cpp-2008 rule
    MISRA_CPP_2008_RULE_0_1_1 = 400010101;
    MISRA_CPP_2008_RULE_0_1_2 = 400010201;
    MISRA_CPP_2008_RULE_0_1_3 = 400010301;
    MISRA_CPP_2008_RULE_0_1_4 = 400010401;
    MISRA_CPP_2008_RULE_0_1_5 = 400010501;
    MISRA_CPP_2008_RULE_0_1_6 = 400010601;
    MISRA_CPP_2008_RULE_0_1_7 = 400010701;
    MISRA_CPP_2008_RULE_0_1_8 = 400010801;
    MISRA_CPP_2008_RULE_0_1_9 = 400010901;
    MISRA_CPP_2008_RULE_0_1_10 = 400011001;
    MISRA_CPP_2008_RULE_0_1_11 = 400011101;
    MISRA_CPP_2008_RULE_0_1_12 = 400011201;
    MISRA_CPP_2008_RULE_0_2_1 = 400020101;
    MISRA_CPP_2008_RULE_0_3_1 = 400030101;
    MISRA_CPP_2008_RULE_0_3_2 = 400030201;
    MISRA_CPP_2008_RULE_0_4_1 = 400040101;
    MISRA_CPP_2008_RULE_0_4_2 = 400040201;
    MISRA_CPP_2008_RULE_0_4_3 = 400040301;
    MISRA_CPP_2008_RULE_1_0_1 = 401000101;
    MISRA_CPP_2008_RULE_1_0_2 = 401000201;
    MISRA_CPP_2008_RULE_1_0_3 = 401000301;
    MISRA_CPP_2008_RULE_2_2_1 = 402020101;
    MISRA_CPP_2008_RULE_2_3_1 = 402030101;
    MISRA_CPP_2008_RULE_2_5_1 = 402050101;
    MISRA_CPP_2008_RULE_2_7_1 = 402070101;
    MISRA_CPP_2008_RULE_2_7_2 = 402070201;
    MISRA_CPP_2008_RULE_2_7_3 = 402070301;
    MISRA_CPP_2008_RULE_2_10_1 = 402100101;
    MISRA_CPP_2008_RULE_2_10_2 = 402100201;
    MISRA_CPP_2008_RULE_2_10_3 = 402100301;
    MISRA_CPP_2008_RULE_2_10_4 = 402100401;
    MISRA_CPP_2008_RULE_2_10_5 = 402100501;
    MISRA_CPP_2008_RULE_2_10_6 = 402100601;
    MISRA_CPP_2008_RULE_2_13_1 = 402130101;
    MISRA_CPP_2008_RULE_2_13_2 = 402130201;
    MISRA_CPP_2008_RULE_2_13_3 = 402130301;
    MISRA_CPP_2008_RULE_2_13_4 = 402130401;
    MISRA_CPP_2008_RULE_2_13_5 = 402130501;
    MISRA_CPP_2008_RULE_3_1_1 = 403010101;
    MISRA_CPP_2008_RULE_3_1_2 = 403010201;
    MISRA_CPP_2008_RULE_3_1_3 = 403010301;
    MISRA_CPP_2008_RULE_3_2_1 = 403020101;
    MISRA_CPP_2008_RULE_3_2_2 = 403020201;
    MISRA_CPP_2008_RULE_3_2_3 = 403020301;
    MISRA_CPP_2008_RULE_3_2_4 = 403020401;
    MISRA_CPP_2008_RULE_3_3_1 = 403030101;
    MISRA_CPP_2008_RULE_3_3_2 = 403030201;
    MISRA_CPP_2008_RULE_3_4_1 = 403040101;
    MISRA_CPP_2008_RULE_3_9_1 = 403090101;
    MISRA_CPP_2008_RULE_3_9_2 = 403090201;
    MISRA_CPP_2008_RULE_3_9_3 = 403090301;
    MISRA_CPP_2008_RULE_4_5_1 = 404050101;
    MISRA_CPP_2008_RULE_4_5_2 = 404050201;
    MISRA_CPP_2008_RULE_4_5_3 = 404050301;
    MISRA_CPP_2008_RULE_4_10_1 = 404100101;
    MISRA_CPP_2008_RULE_4_10_2 = 404100201;
    MISRA_CPP_2008_RULE_5_0_1 = 405000101;
    MISRA_CPP_2008_RULE_5_0_2 = 405000201;
    MISRA_CPP_2008_RULE_5_0_3 = 405000301;
    MISRA_CPP_2008_RULE_5_0_4 = 405000401;
    MISRA_CPP_2008_RULE_5_0_5 = 405000501;
    MISRA_CPP_2008_RULE_5_0_6 = 405000601;
    MISRA_CPP_2008_RULE_5_0_7 = 405000701;
    MISRA_CPP_2008_RULE_5_0_8 = 405000801;
    MISRA_CPP_2008_RULE_5_0_9 = 405000901;
    MISRA_CPP_2008_RULE_5_0_10 = 405001001;
    MISRA_CPP_2008_RULE_5_0_11 = 405001101;
    MISRA_CPP_2008_RULE_5_0_12 = 405001201;
    MISRA_CPP_2008_RULE_5_0_13 = 405001301;
    MISRA_CPP_2008_RULE_5_0_14 = 405001401;
    MISRA_CPP_2008_RULE_5_0_15 = 405001501;
    MISRA_CPP_2008_RULE_5_0_16 = 405001601;
    MISRA_CPP_2008_RULE_5_0_17 = 405001701;
    MISRA_CPP_2008_RULE_5_0_18 = 405001801;
    MISRA_CPP_2008_RULE_5_0_19 = 405001901;
    MISRA_CPP_2008_RULE_5_0_20 = 405002001;
    MISRA_CPP_2008_RULE_5_0_21 = 405002101;
    MISRA_CPP_2008_RULE_5_2_1 = 405020101;
    MISRA_CPP_2008_RULE_5_2_2 = 405020201;
    MISRA_CPP_2008_RULE_5_2_3 = 405020301;
    MISRA_CPP_2008_RULE_5_2_4 = 405020401;
    MISRA_CPP_2008_RULE_5_2_5 = 405020501;
    MISRA_CPP_2008_RULE_5_2_6 = 405020601;
    MISRA_CPP_2008_RULE_5_2_7 = 405020701;
    MISRA_CPP_2008_RULE_5_2_8 = 405020801;
    MISRA_CPP_2008_RULE_5_2_9 = 405020901;
    MISRA_CPP_2008_RULE_5_2_10 = 405021001;
    MISRA_CPP_2008_RULE_5_2_11 = 405021101;
    MISRA_CPP_2008_RULE_5_2_12 = 405021201;
    MISRA_CPP_2008_RULE_5_3_1 = 405030101;
    MISRA_CPP_2008_RULE_5_3_2 = 405030201;
    MISRA_CPP_2008_RULE_5_3_3 = 405030301;
    MISRA_CPP_2008_RULE_5_3_4 = 405030401;
    MISRA_CPP_2008_RULE_5_8_1 = 405080101;
    MISRA_CPP_2008_RULE_5_14_1 = 405140101;
    MISRA_CPP_2008_RULE_5_17_1 = 405170101;
    MISRA_CPP_2008_RULE_5_18_1 = 405180101;
    MISRA_CPP_2008_RULE_5_19_1 = 405190101;
    MISRA_CPP_2008_RULE_6_2_1 = 406020101;
    MISRA_CPP_2008_RULE_6_2_2 = 406020201;
    MISRA_CPP_2008_RULE_6_2_3 = 406020301;
    MISRA_CPP_2008_RULE_6_3_1 = 406030101;
    MISRA_CPP_2008_RULE_6_4_1 = 406040101;
    MISRA_CPP_2008_RULE_6_4_2 = 406040201;
    MISRA_CPP_2008_RULE_6_4_3 = 406040301;
    MISRA_CPP_2008_RULE_6_4_4 = 406040401;
    MISRA_CPP_2008_RULE_6_4_5 = 406040501;
    MISRA_CPP_2008_RULE_6_4_6 = 406040601;
    MISRA_CPP_2008_RULE_6_4_7 = 406040701;
    MISRA_CPP_2008_RULE_6_4_8 = 406040801;
    MISRA_CPP_2008_RULE_6_5_1 = 406050101;
    MISRA_CPP_2008_RULE_6_5_2 = 406050201;
    MISRA_CPP_2008_RULE_6_5_3 = 406050301;
    MISRA_CPP_2008_RULE_6_5_4 = 406050401;
    MISRA_CPP_2008_RULE_6_5_5 = 406050501;
    MISRA_CPP_2008_RULE_6_5_6 = 406050601;
    MISRA_CPP_2008_RULE_6_6_1 = 406060101;
    MISRA_CPP_2008_RULE_6_6_2 = 406060201;
    MISRA_CPP_2008_RULE_6_6_3 = 406060301;
    MISRA_CPP_2008_RULE_6_6_4 = 406060401;
    MISRA_CPP_2008_RULE_6_6_5 = 406060501;
    MISRA_CPP_2008_RULE_7_1_1 = 407010101;
    MISRA_CPP_2008_RULE_7_1_2 = 407010201;
    MISRA_CPP_2008_RULE_7_2_1 = 407020101;
    MISRA_CPP_2008_RULE_7_3_1 = 407030101;
    MISRA_CPP_2008_RULE_7_3_2 = 407030201;
    MISRA_CPP_2008_RULE_7_3_3 = 407030301;
    MISRA_CPP_2008_RULE_7_3_4 = 407030401;
    MISRA_CPP_2008_RULE_7_3_5 = 407030501;
    MISRA_CPP_2008_RULE_7_3_6 = 407030601;
    MISRA_CPP_2008_RULE_7_4_1 = 407040101;
    MISRA_CPP_2008_RULE_7_4_2 = 407040201;
    MISRA_CPP_2008_RULE_7_4_3 = 407040301;
    MISRA_CPP_2008_RULE_7_5_1 = 407050101;
    MISRA_CPP_2008_RULE_7_5_2 = 407050201;
    MISRA_CPP_2008_RULE_7_5_3 = 407050301;
    MISRA_CPP_2008_RULE_7_5_4 = 407050401;
    MISRA_CPP_2008_RULE_8_0_1 = 408000101;
    MISRA_CPP_2008_RULE_8_3_1 = 408030101;
    MISRA_CPP_2008_RULE_8_4_1 = 408040101;
    MISRA_CPP_2008_RULE_8_4_2 = 408040201;
    MISRA_CPP_2008_RULE_8_4_3 = 408040301;
    MISRA_CPP_2008_RULE_8_4_4 = 408040401;
    MISRA_CPP_2008_RULE_8_5_1 = 408050101;
    MISRA_CPP_2008_RULE_8_5_2 = 408050201;
    MISRA_CPP_2008_RULE_8_5_3 = 408050301;
    MISRA_CPP_2008_RULE_9_3_1 = 409030101;
    MISRA_CPP_2008_RULE_9_3_2 = 409030201;
    MISRA_CPP_2008_RULE_9_3_3 = 409030301;
    MISRA_CPP_2008_RULE_9_5_1 = 409050101;
    MISRA_CPP_2008_RULE_9_6_1 = 409060101;
    MISRA_CPP_2008_RULE_9_6_2 = 409060201;
    MISRA_CPP_2008_RULE_9_6_3 = 409060301;
    MISRA_CPP_2008_RULE_9_6_4 = 409060401;
    MISRA_CPP_2008_RULE_10_1_1 = 410010101;
    MISRA_CPP_2008_RULE_10_1_2 = 410010201;
    MISRA_CPP_2008_RULE_10_1_3 = 410010301;
    MISRA_CPP_2008_RULE_10_2_1 = 410020101;
    MISRA_CPP_2008_RULE_10_3_1 = 410030101;
    MISRA_CPP_2008_RULE_10_3_2 = 410030201;
    MISRA_CPP_2008_RULE_10_3_3 = 410030301;
    MISRA_CPP_2008_RULE_11_0_1 = 411000101;
    MISRA_CPP_2008_RULE_12_1_1 = 412010101;
    MISRA_CPP_2008_RULE_12_1_2 = 412010201;
    MISRA_CPP_2008_RULE_12_1_3 = 412010301;
    MISRA_CPP_2008_RULE_12_8_1 = 412080101;
    MISRA_CPP_2008_RULE_12_8_2 = 412080201;
    MISRA_CPP_2008_RULE_14_5_1 = 414050101;
    MISRA_CPP_2008_RULE_14_5_2 = 414050201;
    MISRA_CPP_2008_RULE_14_5_3 = 414050301;
    MISRA_CPP_2008_RULE_14_6_1 = 414060101;
    MISRA_CPP_2008_RULE_14_6_2 = 414060201;
    MISRA_CPP_2008_RULE_14_7_1 = 414070101;
    MISRA_CPP_2008_RULE_14_7_2 = 414070201;
    MISRA_CPP_2008_RULE_14_7_3 = 414070301;
    MISRA_CPP_2008_RULE_14_8_1 = 414080101;
    MISRA_CPP_2008_RULE_14_8_2 = 414080201;
    MISRA_CPP_2008_RULE_15_0_1 = 415000101;
    MISRA_CPP_2008_RULE_15_0_2 = 415000201;
    MISRA_CPP_2008_RULE_15_0_3 = 415000301;
    MISRA_CPP_2008_RULE_15_1_1 = 415010101;
    MISRA_CPP_2008_RULE_15_1_2 = 415010201;
    MISRA_CPP_2008_RULE_15_1_3 = 415010301;
    MISRA_CPP_2008_RULE_15_3_1 = 415030101;
    MISRA_CPP_2008_RULE_15_3_2 = 415030201;
    MISRA_CPP_2008_RULE_15_3_3 = 415030301;
    MISRA_CPP_2008_RULE_15_3_4 = 415030401;
    MISRA_CPP_2008_RULE_15_3_5 = 415030501;
    MISRA_CPP_2008_RULE_15_3_6 = 415030601;
    MISRA_CPP_2008_RULE_15_3_7 = 415030701;
    MISRA_CPP_2008_RULE_15_4_1 = 415040101;
    MISRA_CPP_2008_RULE_15_5_1 = 415050101;
    MISRA_CPP_2008_RULE_15_5_2 = 415050201;
    MISRA_CPP_2008_RULE_15_5_3 = 415050301;
    MISRA_CPP_2008_RULE_16_0_1 = 416000101;
    MISRA_CPP_2008_RULE_16_0_2 = 416000201;
    MISRA_CPP_2008_RULE_16_0_3 = 416000301;
    MISRA_CPP_2008_RULE_16_0_4 = 416000401;
    MISRA_CPP_2008_RULE_16_0_5 = 416000501;
    MISRA_CPP_2008_RULE_16_0_6 = 416000601;
    MISRA_CPP_2008_RULE_16_0_7 = 416000701;
    MISRA_CPP_2008_RULE_16_0_8 = 416000801;
    MISRA_CPP_2008_RULE_16_1_1 = 416010101;
    MISRA_CPP_2008_RULE_16_1_2 = 416010201;
    MISRA_CPP_2008_RULE_16_2_1 = 416020101;
    MISRA_CPP_2008_RULE_16_2_2 = 416020201;
    MISRA_CPP_2008_RULE_16_2_3 = 416020301;
    MISRA_CPP_2008_RULE_16_2_4 = 416020401;
    MISRA_CPP_2008_RULE_16_2_5 = 416020501;
    MISRA_CPP_2008_RULE_16_2_6 = 416020601;
    MISRA_CPP_2008_RULE_16_3_1 = 416030101;
    MISRA_CPP_2008_RULE_16_3_2 = 416030201;
    MISRA_CPP_2008_RULE_16_6_1 = 416060101;
    MISRA_CPP_2008_RULE_17_0_1 = 417000101;
    MISRA_CPP_2008_RULE_17_0_2 = 417000201;
    MISRA_CPP_2008_RULE_17_0_3 = 417000301;
    MISRA_CPP_2008_RULE_17_0_4 = 417000401;
    MISRA_CPP_2008_RULE_17_0_5 = 417000501;
    MISRA_CPP_2008_RULE_18_0_1 = 418000101;
    MISRA_CPP_2008_RULE_18_0_2 = 418000201;
    MISRA_CPP_2008_RULE_18_0_3 = 418000301;
    MISRA_CPP_2008_RULE_18_0_4 = 418000401;
    MISRA_CPP_2008_RULE_18_0_5 = 418000501;
    MISRA_CPP_2008_RULE_18_2_1 = 418020101;
    MISRA_CPP_2008_RULE_18_4_1 = 418040101;
    MISRA_CPP_2008_RULE_18_7_1 = 418070101;
    MISRA_CPP_2008_RULE_19_3_1 = 419030101;
    MISRA_CPP_2008_RULE_27_0_1 = 427000101;
  }
  // for external messages
  ErrorKind error_kind = 6;
  string which_expr = 7;
  string counter_name = 8;
  string ref_name = 9;
  string external_message = 10;
  string filename = 11;
  string other_filename = 12;
  string typedef_decl_name = 13;
  string kind = 14;
  string name = 15;
  string loc = 16;
  string other_loc = 17;
  string tag_name = 18;
  string source_type = 19;
  string destination_type = 20;
  string struct_member_limit = 27;
  string struct_member_count = 28;
  string function_parm_limit = 29;
  string function_parm_count = 30;
  string function_arg_limit = 31;
  string function_arg_count = 32;
  string nested_record_limit = 33;
  string nested_record_count = 34;
  string nested_expr_limit = 35;
  string nested_expr_count = 36;
  string switch_case_limit = 37;
  string switch_case_count = 38;
  string enum_constant_limit = 39;
  string enum_constant_count = 40;
  string string_char_limit = 41;
  string string_char_count = 42;
  string extern_id_limit = 43;
  string extern_id_count = 44;
  string extern_id_char_limit = 45;
  string extern_id_char_count = 46;
  string macro_id_limit = 47;
  string macro_id_count = 48;
  string macro_parm_limit = 49;
  string macro_parm_count = 50;
  string macro_arg_limit = 51;
  string macro_arg_count = 52;
  string nested_block_limit = 53;
  string nested_block_count = 54;
  string nested_include_limit = 55;
  string nested_include_count = 56;
  string iom_id_char_limit = 57;
  string iom_id_char_count = 58;
  string nested_cond_inclu_limit = 59;
  string nested_cond_inclu_count = 60;
  string block_id_limit = 61;
  string block_id_count = 62;
  string nested_decl_limit = 63;
  string nested_decl_count = 64;
  string modify_decl_limit = 65;
  string modify_decl_count = 66;

  int32 severity = 21;
  string code_line_hash = 22;
  string id = 23;
  string language = 24;
  string ruleset = 25;
  string rule_id = 26;
}

// A list of analysis results.
message ResultsList {
  repeated Result results = 1;
}
